File: C/doubly-linked-list/doubly-linked-list.c
#include<stdio.h>

typedef struct{
	struct element* next;
	int key;
	struct element *previous;
}element;


void print_list(element *head) {
	head = head->next;
	while (head != NULL){
		printf("%d->", head->key);
		head = head->next;
	}
	printf("NULL\n");
}

void push_front(element *head, int key) {
	element *tmp = (element*)malloc(sizeof(element));
	tmp->key = key;
	tmp->previous = head;
	tmp->next = head->next;
	
	if(head->next != NULL) ((element *)head->next)->previous = tmp;
	head->next = tmp;
}

void push_back(element *head, int key) {
	element *tmp = (element*)malloc(sizeof(element));
	tmp->key = key;
	tmp->next = NULL;

	while (head->next != NULL) head = head->next;

	head->next = tmp;
	tmp->previous = head;
}

void remove_key(element *head, int key) {
	element *tmp = head->next;

	while (tmp != NULL){
		if (tmp->key == key) {
			if (tmp->previous != NULL) ((element *)tmp->previous)->next = tmp->next;
			if (tmp->next != NULL) ((element *)tmp->next)->previous = tmp->previous;
			
			free(tmp);
			return;
		}
		tmp = tmp->next;
	}
}

int main() {
	element list = { NULL, 0 , NULL };
	push_front(&list, 10);
	push_front(&list, 20);
	push_front(&list, 30);
	push_back(&list, 5);
	push_back(&list, 1);
	print_list(&list);
	remove_key(&list, 30);
	print_list(&list);
	remove_key(&list, 1);
	print_list(&list);
	remove_key(&list, 10);
	remove_key(&list, 100);
	print_list(&list);
	return 0;
}

File: Python/doubly-linked-list/doubly-linked-list.py
class Element:
    def __init__(self, key, next, previous):
        self.key = key
        self.next = next
        self.previous = previous


class List:
    def __init__(self):
        self.head = Element(0, None, None)

    def __str__(self):
        tmp = self.head.next
        object_string = ""
        while tmp is not None:
            object_string += str(tmp.key) + "->"
            tmp = tmp.next
        object_string += "None"
        return object_string

    def push_front(self, key):
        tmp = Element(key, self.head.next, self.head)
        if self.head.next is not None:
            self.head.next.previous = tmp
        self.head.next = tmp

    def push_back(self, key):
        tmp = self.head
        while tmp.next is not None:
            tmp = tmp.next
        tmp.next = Element(key, None, tmp)

    def remove_key(self, key):
        tmp = self.head
        while tmp is not None:
            if tmp.key == key:
                if tmp.previous is not None:
                    tmp.previous.next = tmp.next
                if tmp.next is not None:
                    tmp.next.previous = tmp.previous

                del tmp
                return
            tmp = tmp.next


list = List()
list.push_front(30)
list.push_front(20)
list.push_front(10)
list.push_back(1)
list.push_back(2)
list.push_back(3)
print(list)
list.remove_key(10)
print(list)
list.remove_key(3)
print(list)
list.remove_key(1)
print(list)

